cpp/src/detail/shuffle_wrappers.cu

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cugraph/detail/shuffle_wrappers.hpp>
+#include <cugraph/experimental/detail/graph_utils.cuh>
+#include <cugraph/partition_manager.hpp>
+#include <cugraph/utilities/shuffle_comm.cuh>
+
+#include <raft/random/rng.cuh>
+
+#include <rmm/exec_policy.hpp>
+
+#include <tuple>
+
+namespace cugraph {
+namespace detail {
+
+template <typename vertex_t, typename weight_t>
+std::tuple<rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<vertex_t>,
+           std::optional<rmm::device_uvector<weight_t>>>
+shuffle_edgelist_by_edge(raft::handle_t const& handle,
+                         rmm::device_uvector<vertex_t>& d_edgelist_rows,
+                         rmm::device_uvector<vertex_t>& d_edgelist_cols,
+                         std::optional<rmm::device_uvector<weight_t>>& d_edgelist_weights,
+                         bool store_transposed)
+{
+  auto& comm               = handle.get_comms();
+  auto const comm_size     = comm.get_size();
+  auto& row_comm           = handle.get_subcomm(cugraph::partition_2d::key_naming_t().row_name());
+  auto const row_comm_size = row_comm.get_size();
+  auto& col_comm           = handle.get_subcomm(cugraph::partition_2d::key_naming_t().col_name());
+  auto const col_comm_size = col_comm.get_size();
+
+  // TODO:  Make a shuffle_edges and shuffle_vertices out of these...
+  rmm::device_uvector<vertex_t> d_rx_edgelist_rows(0, handle.get_stream());
+  rmm::device_uvector<vertex_t> d_rx_edgelist_cols(0, handle.get_stream());
+  std::optional<rmm::device_uvector<weight_t>> d_rx_edgelist_weights{std::nullopt};
+  if (d_edgelist_weights) {
+    auto edge_first = thrust::make_zip_iterator(
+      thrust::make_tuple(store_transposed ? d_edgelist_cols.begin() : d_edgelist_rows.begin(),
+                         store_transposed ? d_edgelist_rows.begin() : d_edgelist_cols.begin(),
+                         (*d_edgelist_weights).begin()));
+
+    std::forward_as_tuple(std::tie(store_transposed ? d_rx_edgelist_cols : d_rx_edgelist_rows,
+                                   store_transposed ? d_rx_edgelist_rows : d_rx_edgelist_cols,
+                                   d_rx_edgelist_weights),
+                          std::ignore) =
+      cugraph::experimental::groupby_gpuid_and_shuffle_values(
+        comm,  // handle.get_comms(),
```

Comment:
The comments left here are intentional?

Reply:
I'll clean that up in a separate PR (since this has already merged).  I suspect that was something I did while debugging.